server/py/services/api/initial_data.py

```diff
@@ -983,6 +984,97 @@ def _migrate_data(
+):
+    chunk_size = chunk_size or config.artifacts.artifact_migration_v9_batch_size
+
+    # Query to get latest artifacts that are NOT tagged 'latest'
+    subquery = (
+        db_session.query(
+            framework.db.sqldb.models.ArtifactV2.id,
+            framework.db.sqldb.models.ArtifactV2.key,
+            framework.db.sqldb.models.ArtifactV2.project,
+            framework.db.sqldb.models.ArtifactV2.Tag.name,
+            sqlalchemy.func.row_number()
+            .over(
+                partition_by=[
+                    framework.db.sqldb.models.ArtifactV2.project,
+                    framework.db.sqldb.models.ArtifactV2.key,
+                    framework.db.sqldb.models.ArtifactV2.iteration,
+                ],
+                order_by=framework.db.sqldb.models.ArtifactV2.updated.desc(),
+            )
+            .label("row_number"),
+        )
+        .outerjoin(
+            framework.db.sqldb.models.ArtifactV2.Tag,
+            framework.db.sqldb.models.ArtifactV2.Tag.obj_id
+            == framework.db.sqldb.models.ArtifactV2.id,
+        )
+        .subquery()
+    )
+
+    query = (
+        db_session.query(
+            framework.db.sqldb.models.ArtifactV2.id,
+            framework.db.sqldb.models.ArtifactV2.key,
+            framework.db.sqldb.models.ArtifactV2.project,
+        )
+        .join(subquery, framework.db.sqldb.models.ArtifactV2.id == subquery.c.id)
+        .filter(
+            subquery.c.row_number
+            == 1,  # Only the latest artifact for each project/key/iteration
+            sqlalchemy.or_(
+                subquery.c.name != "latest",  # Exclude if already tagged 'latest'
+                subquery.c.name.is_(None),  # Include artifacts with no tag
+            ),
+        )
+    )
+
+    artifacts_to_tag = query.limit(chunk_size).all()
+
+    if not artifacts_to_tag:
+        logger.info(
+            "No records to migrate", model=framework.db.sqldb.models.ArtifactV2.Tag
+        )
+        return
+
+    logger.info(
+        "Starting migration",
+        model=framework.db.sqldb.models.ArtifactV2.Tag,
+        count=len(artifacts_to_tag),
+    )
+
+    while artifacts_to_tag:
+        new_tags = [
+            framework.db.sqldb.models.ArtifactV2.Tag(
+                obj_id=artifact_id,
+                name="latest",
+                project=project,
+                obj_name=key,
+            )
+            for artifact_id, project, key in artifacts_to_tag
+        ]
+
+        logger.info(
+            "Committing migrated records",
+            model=framework.db.sqldb.models.ArtifactV2.Tag,
+            count=len(new_tags),
+        )
+        db_session.add_all(new_tags)
+        db_session.commit()
+
+        artifacts_to_tag = query.limit(chunk_size).all()
+
+        if not artifacts_to_tag:
+            logger.info(
+                "No more records to migrate",
+                model=framework.db.sqldb.models.ArtifactV2.Tag,
+            )
+            break
```

Comment:
Is there a reason to  use `break` here if the loop will terminate when `artifacts_to_tag` is empty? I see we do it above in  `_ensure_function_kind`  as well, just wondered why